modules/bullet/area_bullet.cpp

```diff
@@ -0,0 +1,281 @@
+	call_event_res[1] = p_otherObject->get_self(); // Other body
+	call_event_res[2] = p_otherObject->get_instance_id(); // instance ID
+	call_event_res[3] = 0; // other_body_shape ID
+	call_event_res[4] = 0; // self_shape ID
+
+	Variant::CallError outResp;
+	areaGodoObject->call(
+			event.event_callback_method, (const Variant **)call_event_res_ptr, 5, outResp);
+}
+
+void AreaBullet::scratch() {
+	if (isScratched)
+		return;
+	isScratched = true;
+}
+
+void AreaBullet::remove_all_overlapping_instantly() {
+	CollisionObjectBullet *supportObject;
+	for (int i = overlappingObjects.size() - 1; 0 <= i; --i) {
+		supportObject = overlappingObjects[i].object;
+		call_event(supportObject, PhysicsServer::AREA_BODY_REMOVED);
+		supportObject->on_exit_area(this);
+	}
+	overlappingObjects.clear();
+}
+
+void AreaBullet::remove_overlapping_instantly(CollisionObjectBullet *p_object) {
+	CollisionObjectBullet *supportObject;
+	for (int i = overlappingObjects.size() - 1; 0 <= i; --i) {
+		supportObject = overlappingObjects[i].object;
+		if (supportObject == p_object) {
+			call_event(supportObject, PhysicsServer::AREA_BODY_REMOVED);
+			supportObject->on_exit_area(this);
+			overlappingObjects.remove(i);
+			break;
+		}
+	}
+}
+
+int AreaBullet::find_overlapping_object(CollisionObjectBullet *p_colObj) {
+	const int size = overlappingObjects.size();
+	for (int i = 0; i < size; ++i) {
+		if (overlappingObjects[i].object == p_colObj) {
+			return i;
+		}
+	}
+	return -1;
+}
+
+void AreaBullet::set_monitorable(bool p_monitorable) {
+	monitorable = p_monitorable;
+}
+
+void AreaBullet::reload_body() {
+	if (space) {
+		space->remove_area(this);
+		space->add_area(this);
+	}
+}
+
+void AreaBullet::set_space(SpaceBullet *p_space) {
+	// Clear the old space if there is one
+	if (space) {
+		isScratched = false;
+
+		// Remove this object form the physics world
+		space->remove_area(this);
+	}
+
+	space = p_space;
+
+	if (space) {
+		space->add_area(this);
+	}
+}
+
+void AreaBullet::on_collision_filters_change() {
+	if (space) {
+		space->reload_collision_filters(this);
+	}
+}
+
+void AreaBullet::add_overlap(CollisionObjectBullet *p_otherObject) {
+	scratch();
+	overlappingObjects.push_back(OverlappingObjectData(p_otherObject, OVERLAP_STATE_ENTER));
+	p_otherObject->notify_new_overlap(this);
+}
+
+void AreaBullet::put_overlap_as_exit(int p_index) {
+	scratch();
+	overlappingObjects[p_index].state = OVERLAP_STATE_EXIT;
+}
+
+void AreaBullet::put_overlap_as_inside(int p_index) {
+	// This check is required to be sure this body was inside
+	if (OVERLAP_STATE_DIRTY == overlappingObjects[p_index].state) {
+		overlappingObjects[p_index].state = OVERLAP_STATE_INSIDE;
+	}
+}
+
+void AreaBullet::set_transform(const Transform &p_global_transform) {
+	btTransform btTrans;
+	G_TO_B(p_global_transform.orthonormalized(), btTrans);
+
+	Vector3 global_ABS_scale;
+	static_cast<const CollisionObjectBullet::HackedBasis &>(p_global_transform.get_basis()).get_ABS_scale(global_ABS_scale);
```

Comment:
You're basically doing the same thing here I explained won't work last week...
Why you have nested rotations and shears in parents and childs, you'll have more than rotation and scaling; you'll have shear, meaning that you can't just split the global transform into a rotation and a scaling.

Reply:
Again, see the explanation and the counter-example I gave on 08-26 on IRC logs.

Reply:
The set transform function is quite equal between Body and Area, but for you is better to work with https://github.com/AndreaCatania/godot/blob/ad3abd23dbb4a7d5550991e2eafa1aacdbbe903d/modules/bullet/body_bullet.cpp

Since we didn't arrive to a solution, What I did after our talk was fix all functions and variables names and implement the code to support shape orientation.

Currently with this code you are able to scale the body and the shape oriented as you wish.

As I've said before Bullet doesn't support shearing and I can set only the translation, rotation, and scaling. As you know the scaling must be set to the shape as local scale.

For this reason my code extrapolate only the information that I need, thus it skip the shearing part, and it converts the body global scale to the shape local scaling.

Now you should have the idea how it works, so can you please improve it?

Reply:
Another thing: As you can see in the code I'm using positive scaling and the negative part was absorbed by rotation

Reply:
OK, to recap: If you skip the shearing that is present in the global transform (which can happen without any shearing in a local transform: for example, if you have a parent with scaling 2,1,1 and a child with rotation 0,0,90), you'll end up with a totally different transform.

What currently happens is, you have a global transform Tg, and you're orthogonalizing it to a matrix O and then getting it's absolute valued scale as S. The problem is, Tg isn't equal to O.S, so you end up with some different transform, leading to different physics.

A composite transform (made of multiple local transforms) cannot be decomposed as a rotation and scaling like Tg = R.S. (except for certain special cases, e.g., when all local scalings in Godot node tree are uniform).

Again, as I tried to explain, you won't have this problem if you don't have a tree like node structure between nodes with transforms, like as in Unity.

And once again, if Bullet doesn't absolutely allow anything other than a global rotation and a local transform, you can't use it with Godot because that's not enough to decompose a composite transform.

I can't fix this problem, this is a limitation in Bullet, and again, maybe Erwin knows a way around this limitation in Bullet API so I'd recommend you to ask his help about this.

Reply:
I've understand your point, if I've not an orthogonal transform I get differences between physics and rendering, since Bullet support only orthogonal transforms. Thanks you!

Reply:
Yeah, the current PR will give incorrect results in general, and there isn't a way to fix the situation on Godot's side (apart from things like forbidding non-uniform scales, or enforcing flat hierarchy...)

So do you have any ideas? Maybe to ask Erwin or modify Bullet?

Reply:
I've talked with reduz and He says that we will show a warning on the editor when a not orthogonal  transform is submitted. So I must care only on orthogonal transforms, I'll implement your advices soon! Thanks!

Reply:
@AndreaCatania The condition for this isn't orthogonality. A single object with scale 2,2,2 will work with the current PR for example, but it's not an [orthogonal matrix](https://en.wikipedia.org/wiki/Orthogonal_matrix).

Reply:
@AndreaCatania I talked to @reduz on IRC, and he's OK not accepting global transforms that can't be decomposed as Bg = Rg.Sg. I'll add a function to Basis to check whether such a decomposition is possible or not (this condition is different from orthogonality, check the counter-example in my previous post). Also, I've reopened #10641 for it, so can you use that function instead?

Reply:
I've updated #10641, the decomposition function also internally checks whether the decomposition is possible or not, and will print a warning message if not (and will return (0,0,0) vector in such cases).